kubeadm: chroot to new --rootfs arg

anguslees opened the pull request:
**What this PR does / why we need it**:

This change adds a new --rootfs=path option to kubeadm, and (if
provided) chroot()s to this path before performing file operations.

This makes it possible to run the kubeadm binary from a container, but
perform remaining file operations against the host filesystem using
something like:

```
    docker run -v /:/rootfs --net=host --uts=host --pid=host \
       kubeadm:latest init ...
```

(Assuming something like the included `examples/kubeadm/Dockerfile` which sets CMD to `kubeadm --rootfs=/rootfs` - Edit: Dockerfile has been removed from this PR, but you get the idea)

Fixes kubernetes/kubeadm#503

**Special notes for your reviewer**:

- I'm not sure where is best to put the Dockerfile, or hook it up to the build process.  Advice sought.

- The kubeadm command line arg handling was less unified than I was expecting to find.  I've implemented this arg for `init` and `join`.   I can add it to all the others too, if we're happy with the approach. An alternative would be to add the arg in the parent `KubeadmCommand`, possibly with a `PersistantFlag` - then it would automatically exist for all kubeadm subcommands.

- It would be slightly preferable if we could order `--rootfs` _before_ the subcommand so we could apply the arg automatically with `ENTRYPOINT ["kubeadm", "--rootfs=/rootfs"]`.  This would be the only such flag in `kubeadm` however, so I have not implemented it that way atm.  (Another alternative would be an env var)

**Release note**:
```release-note
Adds a new EXPERIMENTAL `--rootfs` flag to kubeadm, which (if specified) causes kubeadm to chroot before performing any file operations.  This is expected to be useful when setting up kubernetes on a different filesystem, such as invoking kubeadm from docker.
```

kad commented on cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go:
```diff
@@ -34,6 +34,7 @@ type MasterConfiguration struct {
 	CloudProvider      string     `json:"cloudProvider"`
 	NodeName           string     `json:"nodeName"`
 	AuthorizationModes []string   `json:"authorizationModes,omitempty"`
+	Rootfs             string     `json:"rootfs"`
```
do we really need to expose it here ?

anguslees replied:
Oh I see, I didn't realise how this differed from the other `../types.go`.  No, we don't need it exposed in this (persisted) version.

kad commented on cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go:
```diff
@@ -103,6 +104,7 @@ type NodeConfiguration struct {
 	NodeName                 string   `json:"nodeName"`
 	TLSBootstrapToken        string   `json:"tlsBootstrapToken"`
 	Token                    string   `json:"token"`
+	Rootfs                   string   `json:"rootfs"`
```
same as above.

kad reviewed (commented):
what about `cmd/kubeadm/app/phases/*` ?

kad commented:
For new flags, it would be good also to have release note.

anguslees commented:
@kad: do you want me to _duplicate_ this code for all these subcommands (and phases)?  That seems to be the current pattern, but if I were writing kubeadm I would instead lift this (and perhaps other common options) up to `PersistentFlags` and move some of the runtime logic up to the parent `cobra.Command`'s `Run` method (ie: don't just use `cmdutil.SubCmdRunE` by itself).

I'm happy to do the cut+paste-everywhere version, I just want to be sure that is the direction you want me to go before doing it.

kad commented:
About moving to root `cobra.Command` I have no strong opinion. Most probably it will be better, but I haven't investigated that in details.

My point was purely about functionality: if we introduce functionality in some main commands (join/init),  we should be able to do the same consistently in phases (if kubeadm is used as part of more complex solution which executes phases step-by-step) and in other main commands (reset, upgrade).

php-coder commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -215,6 +220,13 @@ func NewInit(cfgPath string, cfg *kubeadmapi.MasterConfiguration, skipPreFlight,
+		if err := syscall.Chroot(cfg.Rootfs); err != nil {
+			return nil, fmt.Errorf("unable to chroot to %s: %v", cfg.Rootfs, err)
+		}
+		// NB: All file paths after here are relative to Rootfs
```
Note that `chroot` "does not change the current working directory, so that after the call '.' can be outside the tree rooted  at  '/'." See more info in `man 2 chroot`. It's advised to always explicitly invoke `syscall.Chdir` before or after the chroot.

anguslees replied:
Good call - I should not just rely on kubeadm using only absolute paths.  I'll add an explicit chdir(/).

anguslees replied:
(Done)

anguslees commented:
Moved into a `PersistentPreRun` on the top level kubeadm command, so it now automatically applies to all current (and future) subcommands.  It also makes the argument order more suitable for Dockerfile CMD (before the subcommand) - so now the example Dockerfile just automatically applies the `--chroot=/rootfs` arg.

Note that the chroot happens _before_ reading the config file - so the `--config` path is on the host rootfs, not inside the docker image (for the docker use-case example).

anguslees commented:
(I'm going to wait until I get a review, rather than just sit here rebasing this every day.  Please don't misinterpret an outstanding conflict as an indication that I no longer care about this PR.)

kad commented on examples/kubeadm/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Maybe add here as a comment at least some "usage" for this image ? with example of `docker run -it -v /:/rootfs kubeadm init ...` /

anguslees replied:
Done.  Let me know if it doesn't makes sense.

kad commented:
In overall, looks good to me. @anguslees have you thought about looking at creating that kubeadm image as part of release artefacts ? If build scripts will be building this image and populating GCR at the same time as any other release images, it might be easier method for users to get kubeadm.

anguslees commented:
> If build scripts will be building this image and populating GCR at the same time as any other release images, it might be easier method for users to get kubeadm.

Agreed 100% and that is my desire.  As I said in the original PR comment, I have no idea how to go about getting this added/moved to the release images.  Any pointers?  (to docs or people/channels where I can ask again?)

kad commented:
I think first place to look at is: `build/lib/release.sh`, somewhere around `function kube::release::create_docker_images_for_server()`.  But then there is bazel build, and this probably better to ask on #bazel channel in Slack.

errordeveloper commented:
@anguslees is this still of interest? If yes, please rebase.

errordeveloper commented:
Let's get this in, I like the idea and I think it would be useful in LinuxKit (cc @ijc).

ijc commented:
LinuxKit's current model  is to run `kubeadm` in the same container as `kubelet` is running, that container has any necessary binds to the host already in place etc and is naturally where `kubelet` expects to find configuration files anyway, so things mostly seem to work ok. This change would, I suppose, let you run `kubeadm` in a second container if you wanted, which might be useful (although I'm not immediately convinced).

FWIW the implementation looks ok to me, but I'm by no means an expert on `kubeadm`.

kad commented:
@anguslees are you planning to update that PR ?

anguslees commented:
> @anguslees are you planning to update that PR ?

Yes.  I had to wait until the release freeze was over, and by then this had fallen off the top of my todo list :(

Adding a new docker image to the kube release process is also apparently a mildly impossible task unless you happen to be in a very small group of people (and a google employee).  I'll give this another (more serious) attempt.

I'm not going to be able to put time into this again for another 2 weeks (at least).  I'd welcome a PR-to-my-PR that I can just easily merge, if anyone else has the motivation and available cycles...

kad commented:
@anguslees please at least rebase and update kubeadm part

neolit123 commented on cmd/kubeadm/app/util/chroot.go:
```diff
@@ -0,0 +1,44 @@
+
+	if err := syscall.Chroot(rootfs); err != nil {
+		return fmt.Errorf("unable to chroot to %s: %v", rootfs, err)
+	}
```
`syscall.Chroot()` does not exist in the `syscall` package for Windows, so this has to be branched away with `runtime.GOOS != "windows"` if native Windows containers would like to run `kubeadm`.

having said that i don't know the state of the Win32 port.

neolit123 replied:
yep, file suffixes are cleaner even in the case of no-op function for one of the OSes.

anguslees replied:
I added a `chroot_windows.go` which just returns a more descriptive error when `rootfs != ""`

neolit123 commented on cmd/kubeadm/app/util/chroot.go:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	root := filepath.FromSlash("/")
+	if err := os.Chdir(root); err != nil {
```
given the lack of `syscall.Chroot()` on Windows at this point the container would `Chdir` into the host "/", which might not be desired given `kubeadm` would probably be running with elevated privileges.

anguslees replied:
I would think the earlier `syscall.Chroot()` call would return some sort of "not supported" error on any platform without chroot rather than returning silently - so we shouldn't get this far down the function...

neolit123 reviewed (commented):
this is an excellent feature for `*nix` OSes and possibly a compromise can be made for the addition of this flag, given no more command line flags are accepted for kubeadm.

i do not know the state of the Windows port but i do know that there are plans to support the OS. i have added a couple of comments about this and i think this feature should be only enabled on `*nix`.

kad commented:
What should we do with this PR ? We have options to close it or author can rebase with addressing review comments to get it further.

neolit123 commented:
this is very nice to have.
we can potentially cherry pick and co-author / amend the commit in a separate PR.
this will preserve the authorship.

anguslees commented:
I'll see what I can do about rebasing this, and skipping windows cleanly.  I think I should postpone building the actual docker images to a separate PR, because that seems like it involves a separate set of people/approvals. (ie: I'm going to deliberately ignore [this earlier comment](https://github.com/kubernetes/kubernetes/pull/54935#issuecomment-346594204))

neolit123 commented:
@anguslees
my assumption is that this will get approved fast once you have it rebased.
mention me and/or @kad once done.

thanks a lot.

anguslees committed:
kubeadm: chroot to new --rootfs arg

This change adds a new --rootfs=path option to kubeadm, and (if
provided) chroot()s to this path before performing file operations.

This makes it possible to run the kubeadm binary from a container, but
perform remaining file operations against the host filesystem using
something like:

    docker run -v /:/rootfs --net=host --uts=host --pid=host \
       kubeadm:latest init --rootfs /rootfs...

Fixes kubernetes/kubeadm#503

anguslees commented:
/retest

anguslees commented:
@neolit123 ready for your review.

neolit123 commented on cmd/kubeadm/app/util/chroot_unix.go:
```diff
@@ -0,0 +1,46 @@
+// +build !windows
+
+/*
+Copyright 2017 The Kubernetes Authors.
```
2018

anguslees replied:
I wrote the file in 2017 ;)

Updated.

neolit123 commented on cmd/kubeadm/app/util/chroot_windows.go:
```diff
@@ -0,0 +1,34 @@
+		return nil
+	}
+
+	return fmt.Errorf("Chroot feature is unimplemented for windows")
```
`Chroot feature is unimplemented for windows`
->
`chroot is not implemented on Windows`

anguslees replied:
(done)

neolit123 commented on cmd/kubeadm/app/util/chroot_windows.go:
```diff
@@ -0,0 +1,34 @@
+	if rootfs == "" {
+		return nil
+	}
+
```
please remove empty line.

anguslees replied:
(done)

neolit123 commented on cmd/kubeadm/app/util/chroot_unix.go:
```diff
@@ -0,0 +1,46 @@
+	if rootfs == "" {
+		return nil
+	}
+
```
please remove empty line.

anguslees replied:
(done)

neolit123 commented on cmd/kubeadm/app/util/chroot_unix.go:
```diff
@@ -0,0 +1,46 @@
+	if err := syscall.Chroot(rootfs); err != nil {
+		return fmt.Errorf("unable to chroot to %s: %v", rootfs, err)
+	}
+
```
please remove empty line.

anguslees replied:
(done)

neolit123 commented on cmd/kubeadm/app/util/chroot_unix.go:
```diff
@@ -0,0 +1,46 @@
+	if err := os.Chdir(root); err != nil {
+		return fmt.Errorf("unable to chdir to %s: %v", root, err)
+	}
+
```
please remove empty line.

anguslees replied:
(done)

Is this removal of vertical whitespace following a style guide somewhere?  (Is there somewhere I can learn when (not) to add vertical whitespace so I can preemptively avoid future feedback rounds?)

neolit123 commented on cmd/kubeadm/app/cmd/cmd.go:
```diff
@@ -65,6 +69,10 @@ func NewKubeadmCommand(in io.Reader, out, err io.Writer) *cobra.Command {
 		`),
+
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			return kubeadmutil.MaybeChroot(rootfsPath)
```
let's rename `MaybeChroot` to just `Chroot` in symbols and comments.
`Maybe` is not really needed because the function has error handling.

neolit123 reviewed (commented):
hi @anguslees 
thanks for the updat.e

i've added  a couple of small comments - mostly related to cleanup.

neolit123 commented on cmd/kubeadm/app/util/chroot_unix.go:
```diff
@@ -0,0 +1,46 @@
+// NB: All file paths after this call are effectively relative to
+// `rootfs`
+func MaybeChroot(rootfs string) error {
+	if rootfs == "" {
```
also, i think it's a good idea to add a unit test for the unix case of this function.
```
chroot_test.go

// +build !windows
...

func ChrootTest(...)

```
please, let me know if you don't have the time to make these amends.

anguslees replied:
The test will require root (or CAP_SYS_ADMIN) in order to execute chroot().  Are our automated tests run as root? (I hope not..)

neolit123 replied:
that's a good point. they are not!
we shouldn't test the `Chroot()/Chdir()` calls in that case.

but this brings a related topic - `rootfs` is added as a persistent flag for all kubeadm commands, while only `kubead init/join` require root.
i think we should add the flag only for these commands, as something like `kubeadm config ...` or `kubeadm version` would not benefit from that.

WDYT?
also /cc @timothysc

anguslees replied:
I added a test.  Tbh, given it only works when invoked as root, I'm not sure it's worth the additional complexity.

neolit123 commented:
moving these comment outside of the review as it got folded:

>> also, i think it's a good idea to add a unit test

> The test will require root (or CAP_SYS_ADMIN) in order to execute chroot(). Are our automated tests run as root? (I hope not..)

that's a good point. they are not!
we shouldn't test the Chroot()/Chdir() calls in that case.

but this brings a related topic - rootfs is added as a persistent flag for all kubeadm commands, while only kubead init/join require root.
i think we should add the flag only for these commands, as something like kubeadm config ... or kubeadm version would not benefit from that.

WDYT?
also /cc @timothysc

anguslees commented:
> i think we should add the flag only for these commands, as something like kubeadm config ... or kubeadm version would not benefit from that.

I made the flag persistent because [the first review comment](https://github.com/kubernetes/kubernetes/pull/54935#pullrequestreview-73531294) made the point that we need it for all the phases too.

I don't really care either way - afaict this PR has been code complete since November so now I'm just doing whatever the reviewers want :/

Just looking at `config` however, I think it _does_ need to support `--chroot`, since it reads various files like `--kubeconfig`, `--from-file`, `--{old,new}-config`, etc, and we would _want_ those files from the host, not inside the docker image (assuming our original kubeadm-from-docker use case).  Ditto for `token` and other subcommands that use `--kubeconfig`.

I agree `version` has no need for `--chroot`.

As I mentioned [above](https://github.com/kubernetes/kubernetes/pull/54935#issuecomment-341280835), I can duplicate the option in every subcommand other than `version` easily enough, if that's the code structure you'd prefer.

kad commented:
@anguslees I agree, all commands that require reading/writing any file should have `--rootfs` flag to be consistent in expected behaviour. Otherwise this PR looks good for me.

neolit123 commented:
@anguslees @kad 
thank you for having a closer look at the commands that may or many not need `--rootfs`. i also think this PR is good to go now and it's not a big deal that we have commands that don't really benefit from the global flag right now.

thanks for the updates @anguslees and sorry for going back and forth with the reviews and also for this PR collecting dust for so long. i can do any further amends myself if you'd prefer.

/lgtm
pinging @timothysc @luxas @fabriziopandini for review/approval.

neolit123 commented:
/test pull-kubernetes-e2e-kops-aws

anguslees commented:
/test all

anguslees commented:
/retest

neolit123 commented:
/assign @fabriziopandini 
please approve if you get a minute, we can take care of later adjustments to this, but this PR is since Nov 2017 and i think we should get it merged. :\

timothysc reviewed (approved):
This is useful, but there are weird side effects that may occur on chroots.  I'm ok with this so long as --experiment-rootfs is given 

/assign @timothysc

anguslees commented:
/retest

anguslees commented:
> I'm ok with this so long as --experiment-rootfs is given

To be clear: Are you saying the change is ok as-is, or you want me to rename the flag to `--experiment-rootfs`?

(You github-approved the change, but without an "/approve" for the k8s-bots to see)

neolit123 commented:
@anguslees @timothysc 
i guess we can use `--experimental-rootfs`, but if decide to rename the flag this would break the user CLI.
maybe we can keep it `--rootfs`, but instead add the notes that this is **experimental** in both the release note and flag description:

> Add the EXPERIMENTAL `--rootfs` flag to kubeadm, which (if specified) causes kubeadm to chroot before performing any file operations.  This is expected to be useful when setting up kubernetes on a different filesystem, such as invoking kubeadm from docker.

```
"[EXPERIMENTAL] The path to the 'real' host root filesystem.",
```

WDYT?

anguslees commented:
> WDYT?

Happy to type whatever the latest approver wants to see, just need to know what that is.

timothysc commented:
+1 to `"[EXPERIMENTAL] The path to the 'real' host root filesystem.",`

neolit123 commented:
@anguslees ^^

also:
```
Add the EXPERIMENTAL --rootfs...
```
in the release note will be nice to have.
thanks

anguslees committed:
Add "EXPERIMENTAL" to the option description

anguslees commented:
I have prepended `[EXPERIMENTAL]` to the option help text.

neolit123 commented:
@anguslees 
thank you. please squash the commits.

timothysc reviewed (approved):
/lgtm 
/approve

timothysc commented:
/test pull-kubernetes-e2e-kops-aws
